File: Src/Backend/routes/recruitment.py
from flask import Blueprint, jsonify, request
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import Job, JobApplication, User, Notification
from extensions import db
import os
import uuid
from flask import current_app
from utils import get_current_user

recruitment_bp = Blueprint("recruitment_bp", __name__)


@recruitment_bp.route("/jobs", methods=["GET"])
def get_jobs():
    jobs = Job.query.filter_by(status='approved').all()
    output = []
    for job in jobs:
        output.append({
            "id": job.id,
            "title": job.title,
            "company": job.company,
            "location": job.location,
            "salary": job.salary,
            "type": job.type,
            "logo": job.company[0].upper() if job.company else "?",
            "tags": [job.type] if job.type else [],
            "date": job.created_at.strftime("%d/%m/%Y"),
            "description": job.description
        })
    return jsonify({"jobs": output}), 200


@recruitment_bp.route("/jobs/<int:job_id>", methods=["GET"])
def get_job(job_id):
    job = Job.query.get(job_id)
    if not job:
        return jsonify({"msg": "Job not found"}), 404
    return jsonify({
        "id": job.id,
        "title": job.title,
        "company": job.company,
        "location": job.location,
        "salary": job.salary,
        "type": job.type,
        "description": job.description,
        "created_at": job.created_at.isoformat()
    }), 200


@recruitment_bp.route("/jobs", methods=["POST"])
@jwt_required()
def create_job():
    user = get_current_user()

    if user.role not in ["business", "admin"]:
        return jsonify({"msg": "Only businesses can post jobs."}), 403

    data = request.get_json()
    if not data or not data.get("title") or not data.get("company"):
        return jsonify({"msg": "Missing required fields."}), 400

    new_job = Job(
        title=data.get("title"),
        company=data.get("company"),
        location=data.get("location"),
        salary=data.get("salary"),
        type=data.get("type"),
        description=data.get("description"),
        author_id=user.id,
        status='pending'
    )
    db.session.add(new_job)
    db.session.commit()

    return jsonify({"msg": "Job posted successfully. Waiting for admin approval.", "job_id": new_job.id}), 201


@recruitment_bp.route("/jobs/<int:job_id>/apply", methods=["POST"])
@jwt_required()
def apply_job(job_id):
    user = get_current_user()
    job = Job.query.get(job_id)
    if not job:
        return jsonify({"msg": "Job not found"}), 404

    data = request.get_json()
    if not data or not data.get("name") or not data.get("email"):
        return jsonify({"msg": "Missing required fields."}), 400

    existing = JobApplication.query.filter_by(job_id=job_id, user_id=user.id).first()
    if existing:
        return jsonify({"msg": "You have already applied for this job."}), 400

    new_application = JobApplication(
        job_id=job_id,
        user_id=user.id,
        name=data.get("name"),
        email=data.get("email"),
        phone=data.get("phone"),
        cv_url=data.get("cv_url"),
        cover_letter=data.get("cover_letter")
    )
    db.session.add(new_application)
    db.session.commit()

    return jsonify({"msg": "Application submitted successfully."}), 201


@recruitment_bp.route("/applications", methods=["GET"])
@jwt_required()
def get_applications():
    user = get_current_user()

    if user.role == "business":
        jobs = Job.query.filter_by(author_id=user.id).all()
        job_ids = [j.id for j in jobs]
        apps = JobApplication.query.filter(JobApplication.job_id.in_(job_ids)).all()
    else:
        apps = JobApplication.query.filter_by(user_id=user.id).all()

    output = []
    for app in apps:
        output.append({
            "id": app.id,
            "job_title": app.job.title,
            "job_company": app.job.company,
            "name": app.name,
            "email": app.email,
            "phone": app.phone,
            "cv_url": app.cv_url,
            "status": app.status,
            "created_at": app.created_at.isoformat()
        })
    return jsonify({"applications": output}), 200


@recruitment_bp.route("/upload", methods=["POST"])
@jwt_required()
def upload_cv():
    user = get_current_user()

    if 'cv' not in request.files:
        return jsonify({"msg": "No file part"}), 400

    file = request.files['cv']
    if file.filename == '':
        return jsonify({"msg": "No selected file"}), 400

    if file:
        ext = os.path.splitext(file.filename)[1].lower()
        if ext not in ['.pdf', '.doc', '.docx']:
            return jsonify({"msg": "Only PDF, DOC, and DOCX files are allowed."}), 400

        filename = f"cv_{user.id}_{uuid.uuid4().hex}{ext}"
        filepath = os.path.join(current_app.config['UPLOAD_FOLDER'], filename)
        file.save(filepath)

        cv_url = f"/static/uploads/{filename}"
        return jsonify({
            "msg": "CV uploaded successfully",
            "cv_url": cv_url
        }), 200


@recruitment_bp.route("/applications/<int:app_id>/status", methods=["PUT"])
@jwt_required()
def update_application_status(app_id):
    user = get_current_user()

    app = JobApplication.query.get(app_id)
    if not app:
        return jsonify({"msg": "Application not found"}), 404
    if app.job.author_id != user.id and user.role != "admin":
        return jsonify({"msg": "Unauthorized"}), 403

    data = request.get_json()
    status = data.get("status")
    if status not in ['pending', 'reviewed', 'accepted', 'rejected']:
        return jsonify({"msg": "Invalid status"}), 400

    app.status = status
    db.session.commit()
    return jsonify({"msg": f"Application status updated to {status}"}), 200


@recruitment_bp.route("/applications/send-email", methods=["POST"])
@jwt_required()
def send_bulk_email():
    data = request.get_json()
    emails = data.get("emails", [])
    subject = data.get("subject", "Thông báo từ nhà tuyển dụng")
    body = data.get("body", "")

    if not emails:
        return jsonify({"msg": "No emails provided"}), 400

    print(f"Sending email to {len(emails)} recipients: {emails}")
    print(f"Subject: {subject}")

    return jsonify({"msg": f"Successfully sent email to {len(emails)} applicants."}), 200

File: source/backend/routes/recruitment.js
const express = require('express');
const router = express.Router();
const { authenticateToken } = require('../middleware/auth');
const multer = require('multer');
const path = require('path');
const fs = require('fs');

const upload = multer({ dest: 'uploads/' });

// Ensure uploads folder exists
if (!fs.existsSync('uploads')) {
    fs.mkdirSync('uploads');
}

// GET /jobs
router.get('/jobs', async (req, res) => {
    const prisma = req.app.get('prisma');

    try {
        const jobs = await prisma.job.findMany({ where: { status: 'approved' } });
        const output = jobs.map(job => ({
            id: job.id,
            title: job.title,
            company: job.company,
            location: job.location,
            salary: job.salary,
            type: job.type,
            logo: job.company ? job.company[0].toUpperCase() : "?",
            tags: job.type ? [job.type] : [],
            date: job.created_at.toISOString().split('T')[0],
            description: job.description
        }));
        res.status(200).json({ jobs: output });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// GET /jobs/:id
router.get('/jobs/:id', async (req, res) => {
    const prisma = req.app.get('prisma');
    const jobId = parseInt(req.params.id);

    try {
        const job = await prisma.job.findUnique({ where: { id: jobId } });
        if (!job) return res.status(404).json({ msg: "Job not found" });

        res.status(200).json({
            id: job.id,
            title: job.title,
            company: job.company,
            location: job.location,
            salary: job.salary,
            type: job.type,
            description: job.description,
            created_at: job.created_at.toISOString()
        });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// POST /jobs
router.post('/jobs', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);
    const { title, company, location, salary, type, description } = req.body;

    try {
        const user = await prisma.user.findUnique({ where: { id: userId } });
        if (user.role !== 'business' && !user.is_admin) {
            return res.status(403).json({ msg: "Only businesses can post jobs." });
        }

        if (!title || !company) {
            return res.status(400).json({ msg: "Missing required fields." });
        }

        const newJob = await prisma.job.create({
            data: {
                title,
                company,
                location,
                salary,
                type,
                description,
                author_id: userId,
                status: 'pending'
            }
        });

        res.status(201).json({ msg: "Job posted successfully. Waiting for admin approval.", job_id: newJob.id });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// POST /jobs/:id/apply
router.post('/jobs/:id/apply', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);
    const jobId = parseInt(req.params.id);
    const { name, email, phone, cv_url, cover_letter } = req.body;

    try {
        const job = await prisma.job.findUnique({ where: { id: jobId } });
        if (!job) return res.status(404).json({ msg: "Job not found" });

        if (!name || !email) {
            return res.status(400).json({ msg: "Missing required fields." });
        }

        const existing = await prisma.jobApplication.findFirst({
            where: { job_id: jobId, user_id: userId }
        });

        if (existing) {
            return res.status(400).json({ msg: "You have already applied for this job." });
        }

        await prisma.jobApplication.create({
            data: {
                job_id: jobId,
                user_id: userId,
                name,
                email,
                phone,
                cv_url,
                cover_letter
            }
        });

        res.status(201).json({ msg: "Application submitted successfully." });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// GET /applications
router.get('/applications', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);

    try {
        const user = await prisma.user.findUnique({ where: { id: userId } });
        let apps = [];

        if (user.role === 'business') {
            const jobs = await prisma.job.findMany({ where: { author_id: userId } });
            const jobIds = jobs.map(j => j.id);
            apps = await prisma.jobApplication.findMany({
                where: { job_id: { in: jobIds } },
                include: { job: true }
            });
        } else {
            apps = await prisma.jobApplication.findMany({
                where: { user_id: userId },
                include: { job: true }
            });
        }

        const output = apps.map(app => ({
            id: app.id,
            job_title: app.job.title,
            job_company: app.job.company,
            name: app.name,
            email: app.email,
            phone: app.phone,
            cv_url: app.cv_url,
            status: app.status,
            created_at: app.created_at.toISOString()
        }));

        res.status(200).json({ applications: output });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// POST /upload
router.post('/upload', authenticateToken, upload.single('cv'), async (req, res) => {
    const userId = parseInt(req.user.sub);

    if (!req.file) {
        return res.status(400).json({ msg: "No selected file" });
    }

    try {
        const ext = path.extname(req.file.originalname).toLowerCase();
        if (!['.pdf', '.doc', '.docx'].includes(ext)) {
            return res.status(400).json({ msg: "Only PDF, DOC, and DOCX files are allowed." });
        }

        const filename = `cv_${userId}_${Math.random().toString(36).substring(2, 10)}${ext}`;
        const targetPath = path.join(__dirname, '..', 'uploads', filename);

        fs.renameSync(req.file.path, targetPath);

        const cv_url = `/uploads/${filename}`;
        res.status(200).json({ msg: "CV uploaded successfully", cv_url });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// PUT /applications/:id/status
router.put('/applications/:id/status', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);
    const appId = parseInt(req.params.id);
    const { status } = req.body;

    try {
        const app = await prisma.jobApplication.findUnique({
            where: { id: appId },
            include: { job: true }
        });
        if (!app) return res.status(404).json({ msg: "Application not found" });

        const user = await prisma.user.findUnique({ where: { id: userId } });

        if (app.job.author_id !== userId && !user.is_admin) {
            return res.status(403).json({ msg: "Unauthorized" });
        }

        if (!['pending', 'reviewed', 'accepted', 'rejected'].includes(status)) {
            return res.status(400).json({ msg: "Invalid status" });
        }

        await prisma.jobApplication.update({
            where: { id: appId },
            data: { status }
        });

        res.status(200).json({ msg: `Application status updated to ${status}` });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// POST /applications/send-email
router.post('/applications/send-email', authenticateToken, async (req, res) => {
    const { emails, subject, body } = req.body;

    if (!emails || !emails.length) {
        return res.status(400).json({ msg: "No emails provided" });
    }

    console.log(`Sending email to ${emails.length} recipients:`, emails);
    console.log(`Subject: ${subject}`);

    res.status(200).json({ msg: `Successfully sent email to ${emails.length} applicants.` });
});

module.exports = router;
